src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java

```diff
@@ -0,0 +1,248 @@
+        int blockFor = maxBlockFor;
+        for (Replica participant: participants)
+        {
+            // remote dcs can sometimes get involved in dc-local reads. We want to repair
+            // them if they do, but they shouldn't interfere with blocking the client read.
+            if (!repairs.containsKey(participant) && shouldBlockOn(participant.getEndpoint()))
+                blockFor--;
+        }
+
+        // there are some cases where logically identical data can return different digests
+        // For read repair, this would result in ReadRepairHandler being called with a map of
+        // empty mutations. If we'd also speculated on either of the read stages, the number
+        // of empty mutations would be greater than blockFor, causing the latch ctor to throw
+        // an illegal argument exception due to a negative start value. So here we clamp it 0
+        latch = new CountDownLatch(Math.max(blockFor, 0));
+    }
+
+    @VisibleForTesting
+    long waitingOn()
+    {
+        return latch.getCount();
+    }
+
+    @VisibleForTesting
+    boolean isLocal(InetAddressAndPort endpoint)
+    {
+        return ConsistencyLevel.isLocal(endpoint);
+    }
+
+    private boolean shouldBlockOn(InetAddressAndPort endpoint)
+    {
+        return !consistency.isDatacenterLocal() || isLocal(endpoint);
+    }
+
+    @VisibleForTesting
+    void ack(InetAddressAndPort from)
+    {
+        if (shouldBlockOn(from))
+        {
+            // TODO from is NOT Replica!
+            pendingRepairs.remove(from);
+            latch.countDown();
+        }
+    }
+
+    @Override
+    public void response(MessageIn<Object> msg)
+    {
+        ack(msg.from);
+    }
+
+    @Override
+    public boolean isLatencyForSnitch()
+    {
+        return false;
+    }
+
+    private static PartitionUpdate extractUpdate(Mutation mutation)
+    {
+        return Iterables.getOnlyElement(mutation.getPartitionUpdates());
+    }
+
+    /**
+     * Combine the contents of any unacked repair into a single update
+     */
+    private PartitionUpdate mergeUnackedUpdates()
+    {
+        // recombinate the updates
+        List<PartitionUpdate> updates = Lists.newArrayList(Iterables.transform(pendingRepairs.values(), BlockingPartitionRepair::extractUpdate));
+        return updates.isEmpty() ? null : PartitionUpdate.merge(updates);
+    }
+
+    @VisibleForTesting
+    protected void sendRR(MessageOut<Mutation> message, Replica replica)
+    {
+        Preconditions.checkState(replica.isFull(), "cannot send repair mutations to transient replicas");
+        MessagingService.instance().sendRR(message, replica.getEndpoint(), this);
+    }
+
+    public void sendInitialRepairs()
+    {
+        mutationsSentTime = System.nanoTime();
+        for (Map.Entry<Replica, Mutation> entry: pendingRepairs.entrySet())
+        {
+            Replica destination = entry.getKey();
+            Mutation mutation = entry.getValue();
```

Comment:
I'm confused this used to filter out transient replicas, but now it doesn't? Is it happening somewhere else?